### ISSUE TITLE
Make FSDP easier to debug when erroring in backward pass

### ISSUE THREAD

**Issue by rohan-varma** [pytorch/pytorch#83151]
### 🚀 The feature, motivation and pitch

Recently a lot of FSDP enablement efforts have been hard to debug when there's an error in backward pass, because we just get the error "autograd returned null without setting an error" without too much additional detail.

To fix this, we should comb through the code that runs in bwd and ensure we at least use `p_assert` everywhere, and maybe consider more solutions such as wrapping all code that can throw with try/except, adding additional details in the except and throwing a better error.

cc @ezyang @gchanan @zou3519 @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang @kwen2501

**Comment by awgu** [comment]
If I am understanding correctly, only the engine's queued callback should swallow the error on GPU (https://github.com/pytorch/pytorch/issues/75750). Otherwise, we should get error messages properly (e.g. in the various backward hooks).

cc: @albanD

**Comment by albanD** [comment]
Yes. From a quick look at the code we are supposed to catch the error and report it properly here https://github.com/pytorch/pytorch/blob/84d64d72d66b9cccbc58066c731a18bf8493db05/torch/csrc/autograd/engine.cpp#L677
So something fishy is happening with future and it gets lost on the way

**Comment by mori360** [comment]
We are commited to support FSDP2. Here is the migration guide https://github.com/pytorch/torchtitan/blob/main/docs/fsdp.md
